File: src/openfl/filters/BlurFilter.nim
import std/math
import ./BitmapFilter
import ./BitmapFilterQuality

type
  BlurFilter* = ref object of BitmapFilter
    blurXInternal: float64
    blurYInternal: float64
    qualityInternal: int
    horizontalPasses: int
    verticalPasses: int

proc padFor(self: BlurFilter, value: float64): int =
  if value <= 0: return 0
  let passes = if self.qualityInternal > 0: self.qualityInternal else: 1
  let reach = value * float64(passes) * 3.0 # Simplified for Nim port, matching lime logic
  return int(ceil(reach)) + 2

proc blurX*(self: BlurFilter): float64 =
  return self.blurXInternal

proc `blurX=`*(self: BlurFilter, value: float64) =
  if value != self.blurXInternal:
    self.blurXInternal = value
    self.renderDirty = true
    let p = self.padFor(value)
    self.leftExtension = p
    self.rightExtension = p

proc blurY*(self: BlurFilter): float64 =
  return self.blurYInternal

proc `blurY=`*(self: BlurFilter, value: float64) =
  if value != self.blurYInternal:
    self.blurYInternal = value
    self.renderDirty = true
    let p = self.padFor(value)
    self.topExtension = p
    self.bottomExtension = p

proc quality*(self: BlurFilter): int =
  return self.qualityInternal

proc `quality=`*(self: BlurFilter, value: int) =
  self.horizontalPasses = if self.blurXInternal <= 0: 0 else: int(round(self.blurXInternal * (float64(value) / 4.0))) + 1
  self.verticalPasses = if self.blurYInternal <= 0: 0 else: int(round(self.blurYInternal * (float64(value) / 4.0))) + 1
  self.numShaderPasses = self.horizontalPasses + self.verticalPasses

  if value != self.qualityInternal:
    self.renderDirty = true
  self.qualityInternal = value
  self.`blurX=`(self.blurXInternal)
  self.`blurY=`(self.blurYInternal)

proc newBlurFilter*(blurX: float64 = 4, blurY: float64 = 4, quality: int = 1): BlurFilter =
  let self = BlurFilter()
  self.blurXInternal = blurX
  self.blurYInternal = blurY
  self.qualityInternal = quality
  self.needSecondBitmapData = true
  self.preserveObject = false
  self.renderDirty = true
  self.`quality=`(quality)
  return self

method clone*(self: BlurFilter): BitmapFilter =
  return newBlurFilter(self.blurXInternal, self.blurYInternal, self.qualityInternal)

File: kfl/src/main/kotlin/openfl/filters/BlurFilter.kt
package openfl.filters

import openfl.display.BitmapData
import openfl.display.DisplayObjectRenderer
import openfl.display.Shader
import openfl.geom.Point
import openfl.geom.Rectangle
import kotlin.math.ceil
import kotlin.math.pow
import kotlin.math.roundToInt

class BlurFilter(blurX: Double = 4.0, blurY: Double = 4.0, quality: Int = 1) : BitmapFilter() {
    private var __blurX: Double = 0.0
    private var __blurY: Double = 0.0
    private var __horizontalPasses: Int = 0
    private var __quality: Int = 0
    private var __verticalPasses: Int = 0

    var blurX: Double
        get() = __blurX
        set(value) {
            if (value != __blurX) {
                __blurX = value
                __renderDirty = true
                val p = __padFor(value)
                __leftExtension = p
                __rightExtension = p
            }
        }

    var blurY: Double
        get() = __blurY
        set(value) {
            if (value != __blurY) {
                __blurY = value
                __renderDirty = true
                val p = __padFor(value)
                __topExtension = p
                __bottomExtension = p
            }
        }

    var quality: Int
        get() = __quality
        set(value) {
            __horizontalPasses = if (__blurX <= 0) 0 else (Math.round(__blurX * (value.toDouble() / 4.0)) + 1).toInt()
            __verticalPasses = if (__blurY <= 0) 0 else (Math.round(__blurY * (value.toDouble() / 4.0)) + 1).toInt()
            __numShaderPasses = __horizontalPasses + __verticalPasses
            if (value != __quality) __renderDirty = true
            __quality = value
            blurX = __blurX
            blurY = __blurY
        }

    init {
        this.blurX = blurX
        this.blurY = blurY
        this.quality = quality
        __needSecondBitmapData = true
        __preserveObject = false
        __renderDirty = true
    }

    override fun clone(): BitmapFilter {
        return BlurFilter(__blurX, __blurY, __quality)
    }

    override fun __applyFilter(
        bitmapData: BitmapData,
        sourceBitmapData: BitmapData,
        sourceRect: Rectangle,
        destPoint: Point
    ): BitmapData {
        // Skia-based blur would happen here or in the renderer
        return bitmapData
    }

    override fun __initShader(renderer: DisplayObjectRenderer?, pass: Int, sourceBitmapData: BitmapData): Shader? {

        return null
    }

    private fun __padFor(value: Double): Int {
        if (value <= 0) return 0
        val passes = if (__quality > 0) __quality else 1
        val reach = value * passes * 3.0
        return ceil(reach).toInt() + 2
    }
}
